### PR TITLE
fix: permission denied for postgresql

### PR DESCRIPTION
**Pull request by jamesregis**
I implemented the same solution as for awx-operator to solve the problem of permissions on the postgresql folder with an initcontainer.

### CODE REVIEW

**Review by kurokobo — CHANGES_REQUESTED** [review]
Hi, thanks for moving the issue forward.

Have you tested this PR on your side and how?

I haven't done a solid review, just a quick look, but there are some missing default values e.g. `postgres_init_container_commands` and `postgres_data_volume_init` etc.
Also, you've added new params under `database` param, so you can't refer new params unless referring `combined_database` dictionary. As it is, it probably will not work as expected.

If you want to move forward with this PR, please test this PR in your own environment and try to modify the code to make everything to work as expected. Thank you!

**Comment by jamesregis** [comment]
Thanks for comments/review. I'm going to use what you've said to improve the code.

**Comment by jamesregis** [comment]
I have tested his PR before pushing to [jamesregis/eda-server-operator](https://github.com/jamesregis/eda-server-operator) and everything work as expected. postgres start without any manual intervention.

``` 
NAME                                                      READY   STATUS    RESTARTS   AGE
eda-activation-worker-d96cbc94c-9wg4p                     1/1     Running   0          14m
eda-activation-worker-d96cbc94c-d4pfp                     1/1     Running   0          14m
eda-activation-worker-d96cbc94c-tgmf7                     1/1     Running   0          14m
eda-activation-worker-d96cbc94c-vbjbt                     1/1     Running   0          14m
eda-activation-worker-d96cbc94c-zfd22                     1/1     Running   0          14m
eda-api-77788dc6b-7l4cg                                   2/2     Running   0          14m
eda-default-worker-64cc7ff65d-qq7hz                       1/1     Running   0          14m
eda-default-worker-64cc7ff65d-z4jt6                       1/1     Running   0          14m
eda-postgres-15-0                                         1/1     Running   0          16m
eda-redis-747596546f-85qfx                                1/1     Running   0          17m
eda-scheduler-596ccd459-gv42r                             1/1     Running   0          14m
eda-server-operator-controller-manager-7c9b55bdd9-752zs   2/2     Running   0          21m
eda-ui-5c9dbc45d7-7lfqk                                   1/1     Running   0          14m
```

**Review by kurokobo — CHANGES_REQUESTED** [review]
Thanks for your rapid update,

Have you ensured that your PSQL pod has `init` init container? Seems you've reverted changes for CRD so I believe technically no one can specify `postgres_data_volume_init: true` through EDA CR.

Also I think following `if` statement never be true, since there is no `postgres_data_volume_init` variable. Maybe this causes returning undefined variable error. Should use `combined_database.postgres_data_volume_init` instead.

```bash
{% if postgres_data_volume_init and not is_openshift %}
```

So if your PSQL pod has `init` init container, perhaps some change not included in this PR is affecting it. Could you provide the brief steps of your test?

F.Y.I., for a complete test, the following steps should be taken on your branch:

- Build and push custom Operator image: `IMG=<custom_image_name> make docker-build docker-push`
- Apply updated CRDs and deploy Operator: `NAMESPACE=<namespace> IMG=<custom_image_name> make deploy`
- Then deploy EDA CR: `kubectl -n <namespace> apply -f hoge.yaml`
- Ensure expected changes are applied (Not only are pods up and running, but also examine the definition of the pods/statefulset (`kubectl get pod -o yaml`, etc.) to make sure it is what you expect it to be, to ensure original issue has gone or not)

Especially, if some new parameters are added to the CRD as this PR, multiple patterns shuld be tested, at least including a pattern where new parameters are NOT specified in the CR (to test default behavior and side effects), and a pattern where new parameters are specified and their values are changed from default.

If you've done manual test, providing example CR also helps us to review the changes. Thanks!

**Comment by jamesregis** [comment]
I've got it .. let me some changes and I will explain after.

**Comment by jamesregis** [comment]
`postgres_data_volume_init` is now in the CRD under database. User can modify this value by adding `data.postgres_data_volume_init: true`. 

my CR file contains : 
```
---
apiVersion: eda.ansible.com/v1alpha1
kind: EDA
metadata:
  name: eda
  namespace: eda
spec:
  automation_server_url: https://awx.qanope.io
  service_type: ClusterIP
  database:
    postgres_data_volume_init: true 
```
With this file, deployment works as excepted.

```NAME                                                         READY   STATUS    RESTARTS   AGE
pod/eda-activation-worker-d96cbc94c-49286                    1/1     Running   0          7h44m
pod/eda-activation-worker-d96cbc94c-cj974                    1/1     Running   0          7h44m
pod/eda-activation-worker-d96cbc94c-np7bz                    1/1     Running   0          7h44m
pod/eda-activation-worker-d96cbc94c-qfd24                    1/1     Running   0          7h44m
pod/eda-activation-worker-d96cbc94c-zmgjr                    1/1     Running   0          7h44m
pod/eda-api-77788dc6b-zlmh5                                  2/2     Running   0          7h44m
pod/eda-default-worker-64cc7ff65d-7zcwc                      1/1     Running   0          7h44m
pod/eda-default-worker-64cc7ff65d-xcchf                      1/1     Running   0          7h44m
pod/eda-postgres-15-0                                        1/1     Running   0          7h46m
pod/eda-redis-747596546f-x294r                               1/1     Running   0          7h47m
pod/eda-scheduler-596ccd459-t4td8                            1/1     Running   0          7h44m
pod/eda-server-operator-controller-manager-5446d478b-9lprp   2/2     Running   0          7h49m
pod/eda-ui-5c9dbc45d7-z7dlh                                  1/1     Running   0          7h44m



NAME                                                             TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
service/eda-api                                                  ClusterIP   10.233.9.176    <none>        8000/TCP   7h45m
service/eda-daphne                                               ClusterIP   10.233.23.163   <none>        8001/TCP   7h45m
service/eda-postgres-15                                          ClusterIP   None            <none>        5432/TCP   7h46m
service/eda-redis-svc                                            ClusterIP   10.233.36.44    <none>        6379/TCP   7h47m
service/eda-server-operator-controller-manager-metrics-service   ClusterIP   10.233.31.222   <none>        8443/TCP   7h49m
service/eda-ui                                                   ClusterIP   10.233.13.74    <none>        80/TCP     7h44m

NAME                                                     READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/eda-activation-worker                    5/5     5            5           7h44m
deployment.apps/eda-api                                  1/1     1            1           7h44m
deployment.apps/eda-default-worker                       2/2     2            2           7h44m
deployment.apps/eda-redis                                1/1     1            1           7h47m
deployment.apps/eda-scheduler                            1/1     1            1           7h44m
deployment.apps/eda-server-operator-controller-manager   1/1     1            1           7h49m
deployment.apps/eda-ui                                   1/1     1            1           7h44m

NAME                                                               DESIRED   CURRENT   READY   AGE
replicaset.apps/eda-activation-worker-d96cbc94c                    5         5         5       7h44m
replicaset.apps/eda-api-77788dc6b                                  1         1         1       7h44m
replicaset.apps/eda-default-worker-64cc7ff65d                      2         2         2       7h44m
replicaset.apps/eda-redis-747596546f                               1         1         1       7h47m
replicaset.apps/eda-scheduler-596ccd459                            1         1         1       7h44m
replicaset.apps/eda-server-operator-controller-manager-5446d478b   1         1         1       7h49m
replicaset.apps/eda-ui-5c9dbc45d7                                  1         1         1       7h44m

NAME                               READY   AGE
statefulset.apps/eda-postgres-15   1/1     7h46m
```

Postgresql pod contains:

```
.....
  hostname: eda-postgres-15-0
  initContainers:
  - command:
    - /bin/sh
    - -c
    - |
      chown 26:0 /var/lib/pgsql/data
      chmod 700 /var/lib/pgsql/data
    image: quay.io/sclorg/postgresql-15-c9s:latest
    imagePullPolicy: IfNotPresent
    name: init
    resources:
      requests:
        cpu: 50m
        memory: 100Mi
    securityContext:
      runAsUser: 0
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/lib/pgsql/data
      name: postgres-15
      subPath: data
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-zzcj5
      readOnly: true
.....
```
When `data.postgres_data_volume_init` is not set, default value is `false` in the CRD, I've got a database `permission error`.

Let me know if I' ve done something wrong.

**Comment by kurokobo** [comment]
@jamesregis 
Thanks for updating again, everything seems great! Also thanks for explaining the testing methods and results, very clear.

I also have tested this PR and I can confirm this works as expected. LGTM 😃 

Your contribution is significant even up to this point, but if you are interested, why don't you update the document as well?
I think updating this file is good: https://github.com/ansible/eda-server-operator/blob/main/docs/user-guide/database-configuration.md
You can refer to the "Initialize Postgres data volume" section in AWX Operator docs as a reference: https://ansible.readthedocs.io/projects/awx-operator/en/latest/user-guide/database-configuration.html#initialize-postgres-data-volume

Of cource I (or any other contributer) can update docs by follow-up PR, but I want to respect your accomplishments as much as possible. Thanks!

@rooftopcellist 
I don't make any test on OpenShift, so if it is required to ensure there is no side-effect on OpenShift, please make additional test. Thanks!

**Comment by jamesregis** [comment]
@kurokobo, 
I will update the documentation then to complete the PR. 😃

File: docs/user-guide/database-configuration.md
Comment (kurokobo) [inline docs/user-guide/database-configuration.md:63]: `database.` instead of `data.`

File: docs/user-guide/database-configuration.md
Comment (kurokobo) [inline docs/user-guide/database-configuration.md:101]:
```suggestion
To initialize this directory with the correct permissions, add `database.postgres_data_volume_init: true` to EDA instance.
```

File: docs/user-guide/database-configuration.md
Comment (kurokobo) [inline docs/user-guide/database-configuration.md:114]: I think a minimal example is sufficient. Also it's helpful if we have an example of `database.postgres_init_container_commands`

````suggestion

```yaml
spec:
  database:
    postgres_data_volume_init: true
```

Should you need to modify the init container commands, there is an example below.

```yaml
spec:
  database:
    postgres_data_volume_init: true
    postgres_init_container_commands: |
      chown 26:0 /var/lib/pgsql/data
      chmod 700 /var/lib/pgsql/data
```
````

**Review by kurokobo — CHANGES_REQUESTED** [review]
Hi, thanks for updating the docs! Added some comments about minor issues. Thanks!

**Comment by jamesregis** [comment]
@kurokobo thanks for taking the time to help me with my first commit on eda-server-operator. I hope I can make some more in the future ! :smile:

File: docs/user-guide/database-configuration.md
Comment (kurokobo) [inline docs/user-guide/database-configuration.md:63]: Oh you've forgotten fix here 😃 

```suggestion
| database.postgres_data_volume_init                |  Initialize PostgreSQL data directory with the correct permissions | false |
```

**Review by kurokobo — APPROVED** [review]
@jamesregis 
Thanks for all your work.
I'm not a maintainer but just a contributor so I can't officially approve this, but I want to say this is LGTM both codes and docs 😃
Please wait for the review from maintainers. Thanks!

**Comment by jamesregis** [comment]
> @jamesregis Thanks for all your work. I'm not a maintainer but just a contributor so I can't officially approve this, but I want to say this is LGTM both codes and docs 😃 Please wait for the review from maintainers. Thanks!

Thanks a lot @kurokobo, your help was very helpful helping me produce a better code for the operator :+1:

**Review by rooftopcellist — APPROVED** [review]

**Comment by rooftopcellist** [comment]
Thank you both for all of the time and effort you put in to implementing and testing this change.  I deployed with this code and couldn't find any issues.  

Thanks!

**Pull request merged**
